src/main/java/org/deepsymmetry/beatlink/dbserver/ConnectionManager.java

```diff
@@ -280,59 +281,76 @@ private void setPlayerDBServerPort(int player, int port) {
+                    socket.connect(address, socketTimeout.get());
+                    InputStream is = socket.getInputStream();
+                    OutputStream os = socket.getOutputStream();
+                    socket.setSoTimeout(socketTimeout.get());
+                    os.write(DB_SERVER_QUERY_PACKET);
+                    byte[] response = readResponseWithExpectedSize(is);
+                    if (response.length == 2) {
+                        final int portReturned = (int)Util.bytesToNumber(response, 0, 2);
+                        if (portReturned == 65535) {
+                            logger.info("Player {} reported dbserver port of {}, not yet ready?", announcement.getDeviceNumber(), portReturned);
+                        } else {
+                            if (isRunning()) {  // Bail if we were shut down before we received a response.
+                                dbServerPorts.put(announcement.getAddress(), portReturned);
+                            }
+                            return;  // Success!
+                        }
                     }
-                }
-            } catch (java.net.ConnectException ce) {
-                logger.info("Player {} doesn't answer rekordbox port queries, connection refused, not yet ready?", announcement.getDeviceNumber());
-            } catch (Throwable t) {
-                logger.warn("Problem requesting database server port number", t);
-            } finally {
-                if (socket != null) {
-                    try {
-                        socket.close();
-                    } catch (IOException e) {
-                        logger.warn("Problem closing database server port request socket", e);
+                } catch (java.net.ConnectException ce) {
+                    logger.info("Player {} doesn't answer rekordbox port queries, connection refused, not yet ready?", announcement.getDeviceNumber());
+                } catch (Throwable t) {
+                    logger.warn("Problem requesting database server port number", t);
+                } finally {
+                    if (socket != null) {
+                        try {
+                            socket.close();
+                        } catch (IOException e) {
+                            logger.warn("Problem closing database server port request socket", e);
+                        }
                     }
                 }
             }
-        }
 
-        logger.info("Player {} never responded with a valid rekordbox dbserver port. Won't attempt to request metadata.",
-                announcement.getDeviceNumber());
+            logger.info("Player {} never responded with a valid rekordbox dbserver port. Won't attempt to request metadata.",
+                    announcement.getDeviceNumber());
+        } catch (Throwable t) {
+            logger.error("Problem querying for database server port on device {}:", announcement.getDeviceNumber(), t);
+        } finally {
```

Comment:
This new finally clause is critical. It is here that we remove the thread from the map of addresses to active threads querying them; no matter how the thread ends, via normal return, early return, or exception, this code will run, and remove it from that map, so we can know it is no longer active.